docs/source/best-practices/checkpointing.rst

```diff
@@ -0,0 +1,162 @@
+
+
+Checkpointing on Discovery 
+================================
+
+Checkpointing will allow you to: 
+
+ * Create resilient workflows in the existence of faults.
+ * Overcome most scheduler resource time limitations.
+ * Implement an early error detection approach by inspecting intermediate results. 
+
+Checkpointing can be acheived by a combination of implementing Checkpointing at some level in your job with the use of `Slurm Job Arrays <https://slurm.schedmd.com/job_array.html>`_. 
+
+.. note::
+   To overcome partition time limits, replace your long single job with multiple shorter jobs. Using job arrays, set each job to run after another. Each job will write a checkpoint file if checkpointing is implemented. The next job in line will the latest checkpoint file to continue from the latest state of the calculation.
+
+GROMACS Checkpointing example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example demonstrates how to implement a longer `GROMACS <https://www.gromacs.org/>`_ 120 hour calculation on using multiple shorter jobs on the **short** partition with Slurm job arrays and the GROMACS built-in Checkpointing option (read more `here <https://manual.gromacs.org/documentation/current/user-guide/managing-simulations.html>`_).
+
+The following script **submit_mdrun_array.bash** creates a Slurm job array of ::
+
+ #!/bin/bash
+ #SBATCH --partition=short
+ #SBATCH --constraint=cascadelake
+ #SBATCH --nodes=1
+ #SBATCH --time=12:00:00
+ #SBATCH --job-name=myrun
+ #SBATCH --ntasks=56
+ #SBATCH --array=1-10%1  #execute 10 array jobs, 1 at a time.
+ #SBATCH --output=myrun-%A_%a.out
+ #SBATCH --error=myrun-%A_%a.err
+ 
+ module load cuda/10.2
+ module load gcc/7.3.0
+ module load openmpi/4.0.5-skylake-gcc7.3
+ module load gromacs/2020.3-gpu-mpi
+ source /shared/centos7/gromacs/2020.3-gcc7.3/bin/GMXRC.bash
+
+ srun --mpi=pmi2 -n $SLURM_NTASKS gmx_mpi mdrun -ntomp 1 -s myrun.tpr -v -dlb yes -cpi state
+
+Where we used the checkpoint flag followed by the file name `-cpi state` to be used to checkpointing. This directs mdrun to use the checkpoint file named `state.cpt` when loading the state. The Slurm option `--array=1-10%1` will create 10 Slurm array tasks, and will run one task job at a time for 12 hours. Note that the saved variable `%A` denotes the main job ID, while variable `%a` denotes the task ID (spanning values 1-10).
+
+To submit these jobs to the scheduler, use the command::
+
+   sbatch submit_mdrun_array.bash
+
+Python TensorFlow Checkpointing example
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This example demonstrates how implement a longer TensorFlow ML training using the **tf.keras** Checkpointing `API <https://www.tensorflow.org/tutorials/keras/save_and_load>`_ and multiple shorter Slurm job arrays on the gpu partition.
```

Comment:
...demonstrates how **to** implement a longer TensorFlow ML training **job** or **task**...